crates/ra_hir/src/expr.rs

```diff
@@ -56,13 +56,14 @@ pub struct Body {
 /// file, so that we don't recompute types whenever some whitespace is typed.
 #[derive(Default, Debug, Eq, PartialEq)]
 pub struct BodySourceMap {
-    expr_map: FxHashMap<SyntaxNodePtr, ExprId>,
-    expr_map_back: ArenaMap<ExprId, SyntaxNodePtr>,
+    expr_map: FxHashMap<ExprPtr, ExprId>,
```

Comment:
this is probably the most interesting change here. Though it seems like in the future we should replace this `Either` with something less ad-hoc: I imagine we might explicitelly store `is_deshugarded: true` flag here.